rust/pattern/constraint/type_constraint.rs

```diff
@@ -0,0 +1,273 @@
+impl From<String> for RelatesConstraint {
+    fn from(type_name: String) -> Self {
+        RelatesConstraint {
+            role_type: UnboundVariable::hidden()
+                .type_(ScopedType::from((String::from("relation"), type_name)))
+                .unwrap()
+                .into_type(),
+            overridden_role_type: None,
+        }
+    }
+}
+
+impl From<UnboundVariable> for RelatesConstraint {
+    fn from(role_type: UnboundVariable) -> Self {
+        RelatesConstraint::from(role_type.into_type())
+    }
+}
+impl From<TypeVariable> for RelatesConstraint {
+    fn from(role_type: TypeVariable) -> Self {
+        RelatesConstraint { role_type, overridden_role_type: None }
+    }
+}
+
+impl Display for RelatesConstraint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "relates {}", self.role_type)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct PlaysConstraint {
+    pub role_type: TypeVariable,
+    pub relation_type: Option<TypeVariable>,
+    pub overridden_role_type: Option<TypeVariable>,
+}
+
+impl PlaysConstraint {
+    fn new(role_type: TypeVariable, overridden_role_type: Option<TypeVariable>) -> Self {
+        PlaysConstraint {
+            relation_type: role_type.label.as_ref().map(|label| {
+                UnboundVariable::hidden()
+                    .type_(label.scoped_type.scope.as_ref().cloned().unwrap())
+                    .unwrap()
+                    .into_type()
+            }),
+            role_type,
+            overridden_role_type,
+        }
+    }
+}
+
+impl From<(&str, &str)> for PlaysConstraint {
+    fn from((relation_type, role_type): (&str, &str)) -> Self {
+        PlaysConstraint::from((String::from(relation_type), String::from(role_type)))
+    }
+}
+
+impl From<(String, String)> for PlaysConstraint {
+    fn from((relation_type, role_type): (String, String)) -> Self {
+        PlaysConstraint::from(ScopedType::from((relation_type, role_type)))
+    }
+}
+
+impl From<ScopedType> for PlaysConstraint {
+    fn from(scoped_type: ScopedType) -> Self {
+        PlaysConstraint::new(
+            UnboundVariable::hidden().type_(scoped_type).unwrap().into_type(),
+            None,
+        )
+    }
+}
+
+impl From<UnboundVariable> for PlaysConstraint {
+    fn from(role_type: UnboundVariable) -> Self {
+        PlaysConstraint::from(role_type.into_type())
+    }
+}
+impl From<TypeVariable> for PlaysConstraint {
+    fn from(role_type: TypeVariable) -> Self {
+        PlaysConstraint::new(role_type, None)
+    }
+}
+
+impl Display for PlaysConstraint {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "plays {}", self.role_type)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum IsKey {
+    Yes,
+    No,
+}
+
+impl Display for IsKey {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if *self == IsKey::Yes {
+            f.write_str(" @key")?;
+        }
+        Ok(())
+    }
+}
```

Comment:
Enums are generally better than naked bools for call site readability. Compare `owns(("name", true))` and `owns(("name", IsKey::Yes))`. Might be nice to have a global constant, say, `KEY`, so that we could write `owns(("name", KEY))` instead.